client/taskenv/env.go

```diff
@@ -476,15 +476,25 @@ func (b *Builder) Build() *TaskEnv {
 	// Clean keys (see #2405)
 	cleanedEnv := make(map[string]string, len(envMap))
 	for k, v := range envMap {
-		cleanedK := helper.CleanEnvVar(k, '_')
+		var cleanedK string
+		if strings.HasPrefix(k, "NOMAD_") {
+			cleanedK = helper.CleanEnvVar(k, '_')
```

Comment:
Which contexts does nomad generate environment variables with names containing "unclean" characters?  It may make sense to clean those only on setting instead of here.  I suspect the only cases are label names in https://github.com/hashicorp/nomad/blob/4d709ee3b41cd9009a49fad0f970a505a1a9e875/client/taskenv/env.go#L730-L740 .  Other nomad generated env-vars names are hard-coded.

Handling it at setting time allows us to treat user set environment variables consistently even if user prefixed it with `NOMAD_`.  The `NOMAD_` prefix check is a imperfect heuristic.